text/0566-memo-decorator.md

```diff
@@ -0,0 +1,141 @@
+- Start Date: 2019-12-22
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/566
+- Tracking: (leave this empty)
+
+# @memo
+
+## Summary
+
+Add a `@memo` decorator for memoizing the result of a getter based on
+autotracking. In the following example, `fullName` would only recalculate if
+`firstName` or `lastName` is updated.
+
+```js
+class Person {
+  @tracked firstName = 'Jen';
+  @tracked lastName = 'Weber';
+
+  @memo
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+## Motivation
+
+One of the major differences between computed properties and tracked properties
+with autotracking in Octane is that native, autotracked getters do not
+automatically cache their values, where computed properties were cached by
+default. This was an intentional design choice, as the memoization logic for
+computed properties was actually more costly, on average, than rerunning the
+getter in the first place. This was especially true given that computed
+properties would usually only ever be calculated and used once or twice per
+render before being updated.
+
+However, there are absolutely cases where getters _are_ expensive, and their
+values are used repeatedly, so memoization would be very helpful. Strategic,
+opt-in memoization is a useful tool that would help Ember developers optimize
+their apps when relevant, without adding extra overhead unless necessary.
+
+## Detailed design
+
+The `@memo` decorator will be exported from `@glimmer/tracking`, alongside
+`@tracked`. It can be used on native getters to memoize their return values
+based on the tracked state they consume while being calculated.
+
+```js
+import { tracked, memo } from '@glimmer/tracking';
+
+class Person {
+  @tracked firstName = 'Jen';
+  @tracked lastName = 'Weber';
+
+  @memo
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+In this example, the `fullName` getter will be memoized whenever it is called,
+and will only be recalculated the next time the `firstName` or `lastName`
+properties are set. This would apply to any autotracking tags consumed while
+calculating the getter, so changes to `EmberArray`s and other tracked
+primitives, for instance, would also cause invalidations.
```

Comment:
That’s exactly what that was meant to capture. Essentially, any autotrack-able change should bust the cache. If it would invalidate the template, it would invalidate `@memo`.